### PR TITLE
fix: various failures and issues

### LINKED ISSUE #177: Analyze Groovy files in folder fails
If you select the option to Analyze Groovy files in folder from the command pallet it fails as there's no folder selected.

```log
[Error - 2:14:30 AM] Request workspace/executeCommand failed.
  Message: Request workspace/executeCommand failed with message: Cannot read properties of undefined (reading 'map')
  Code: -32603 
```

### PR DESCRIPTION
**Pull request by stevenh**
Refactor tests to be more reliable, making them independent of each other, so if one test fails others are not effected. This includes self managing timeouts as mocha doesn't handle pending promises, which makes it harder to debug issues.

Run and Debug commands:
* Fix path to workspace files of "Language Server E2E Test" Run and Debug command.
* Rename to make it clear what each config is used for.

Add mochaExplorer configuration, to improve test visibility.

Re-enable big groovy test file, now it works as expected.

Rename test files to eliminate stuttering.

Fix test debugging by using NPM_DEBUG instead of DEBUG env var which is filtered out by vscode, see: https://github.com/microsoft/vscode/issues/197494.

Remove duplicate lint call that's now handled by onDidChangeContent.

Remove fixed delay comment so we don't have to update when the delay changes.

Update Node and JVM versions in README.md.

Update all packages, to address security issues and bring in the latest version of npm-groovy-lint which includes a critical race condition fix.

Clean up imports.

Use onDidChangeContent to trigger re-linting to improve performance and remove skipNextOnDidChangeContents which
conflicts with this change.

Eliminate else statements where previous block returns, to improve code readability.

Refactor resetDiagnostics into deleteDiagnostics as its nowonly used for the delete flow, so we can remove unnecessary
logic and diagnostic changes.

Remove commented out code.

Convert high noise debugging to trace.

Add missing await to docManager.updateDocumentSettings call

Use latest xvfb-action@v1 instead of v1.0 to bring in the latest fixes.

Fixes #177 #174 #181 #172

### Author/Contributor
- [x] Add entry to the **CHANGELOG.md** listing the change and linking to the corresponding issue (if appropriate)
- [x] If documentation is needed for this change, has that been included in this pull request

### Reviewing Maintainer
- [ ] Label as `breaking` if this is a large fundamental change
- [ ] Label as either `automation`, `bug`, `documentation`, `enhancement`, `infrastructure`, or `performance`

### CODE REVIEW

File: client/package.json [hunk 1]
```diff
@@ -23,20 +23,24 @@
...
+    "vscode": "^1.75.0",
     "node": ">=14.0.0"
   },
   "author": "Nicolas Vuillamy",
   "license": "GPL-3.0-only",
   "dependencies": {
-    "vscode-languageclient": "^6.1.1"
+    "@types/vscode": "^1.85.0",
+    "glob": "^10.3.10",
+    "typescript": "^5.3.3",
```
Comment (nvuillam) [inline client/package.json:35]: Are you sure you need the full typescript dependency here, as the generated extension VSIX is in JS ?

Comment (nvuillam) [reply client/package.json:35]: Maybe more in devDependencies N

Comment (nvuillam) [reply client/package.json:35]: Maybe "@types/vscode" is devDependencies too ? :)

File: server/package.json [hunk 2]
```diff
@@ -30,16 +30,20 @@
...
   "dependencies": {
-    "debug": "^4.1.1",
-    "find-java-home": "^1.1.0",
-    "fs-extra": "^8.1.0",
-    "glob-promise": "^3.4.0",
-    "npm-groovy-lint": "^10.0.0",
-    "vscode-languageserver": "^6.1.1",
-    "vscode-languageserver-textdocument": "^1.0.1",
-    "vscode-uri": "^2.1.1"
+    "@types/vscode": "^1.85.0",
```
Comment (nvuillam) [inline server/package.json:33]: shoudn't @types/vscode be in devDependencies ?

**Comment by nvuillam** [comment]
@stevenh it's like you deep dived in my mind in the past, understood everything and made it more elegant, that's amazing, can't wait to see this PR merged and build a new golden statue :)

**Comment by nvuillam** [comment]
I see the test cases... they seem to have long execution time, maybe CodeNarc Server is not called and npm-groovy-lint fallbacks to new java command line ? :/

**Comment by stevenh** [comment]
> I see the test cases... they seem to have long execution time, maybe CodeNarc Server is not called and npm-groovy-lint fallbacks to new java command line ? :/

There's definitely some slowness, focusing in on stability for now.

File: server/package.json (hunk 2 above)
Comment (stevenh) [reply server/package.json:33]: Quite likely, there was a lot of messing around during the development of this to get things playing nice, need to do a self review and clean up pass.

File: client/package.json (hunk 1 above)
Comment (stevenh) [reply client/package.json:35]: > Are you sure you need the full typescript dependency here, as the generated extension VSIX is in JS ?

Will check on this, thanks or flagging.

**Comment by stevenh** [comment]
Looks like there is still a race deep in the code, it's not clear if its vscode extension, npm-groovy-lint or both, but still more work to do.

**Comment by stevenh** [comment]
Main race is fixed https://github.com/nvuillam/npm-groovy-lint/pull/343

**Comment by stevenh** [comment]
@nvuillam apologies for another huge set of changes, but wanted to get to a place where the tests were solid, which has take quite a refactor.

I think we're finally there now, which should give us a solid base to move forward.

Please pay particular attention to the logic changes in the server component, when reviewing, as there could be some reason for the original flow which I didn't understand.

**Comment by nvuillam** [comment]
@stevenh the original flow was to cancel the current linting request when there is a new one incoming for the same file ^^ ( lint while wiring code)

Is such feature still handled ? :)

**Review by nvuillam — APPROVED** [review]
Amazing work :)
I cloned your branch locally and made a few successful tests :)
Time for a new golden statue :)

**Comment by stevenh** [comment]
> @stevenh the original flow was to cancel the current linting request when there is a new one incoming for the same file ^^ ( lint while wiring code)
> 
> Is such feature still handled ? :)

Thanks for that context, yes that case should still be handled 😀

**Comment by nvuillam** [comment]
@stevenh by the way... sorry, I used Canva.com again 🤣 

image

**Pull request merged**